gemma-web/src/main/java/ubic/gemma/web/controller/GeneralSearchControllerImpl.java

```diff
@@ -282,122 +267,6 @@ protected ModelAndView showForm( HttpServletRequest request, HttpServletResponse
-                cvos.add( ch );
-            }
-            vos = cvos;
-        } else if ( BioSequenceValueObject.class.isAssignableFrom( entityClass ) ) {
-            return; // why?
-        } else if ( GeneSet.class.isAssignableFrom( entityClass ) ) {
-            vos = geneSetService.getValueObjects( ids );
-        } else if ( ExpressionExperimentSet.class.isAssignableFrom( entityClass ) ) {
-            vos = experimentSetService.loadValueObjects( experimentSetService.load( ids ) );
-        } else if ( BlacklistedEntity.class.isAssignableFrom( entityClass ) ) {
-            Collection<BlacklistedValueObject> bvos = new ArrayList<>();
-            for ( SearchResult sr : results ) {
-                bvos.add( BlacklistedValueObject.fromEntity( ( BlacklistedEntity ) sr.getResultObject() ) );
-            }
-            vos = bvos;
-        } else {
-            throw new UnsupportedOperationException( "Don't know how to make value objects for class=" + entityClass );
-        }
-
-        if ( vos == null || vos.isEmpty() ) {
-
-            // bug 3475: if there are search results but they are all removed because they are troubled, then results
-            // has ExpressionExperiments in
-            // it causing front end errors, if vos is empty make sure to get rid of all search results
-            for ( Iterator<SearchResult<?>> it = results.iterator(); it.hasNext(); ) {
-                it.next();
-                it.remove();
-            }
-
-            return;
-        }
-
-        // retained objects...
-        Map<Long, Identifiable> idMap = EntityUtils.getIdMap( vos );
-
-        for ( Iterator<SearchResult<?>> it = results.iterator(); it.hasNext(); ) {
-            SearchResult sr = it.next();
-            if ( !idMap.containsKey( sr.getResultId() ) ) {
-                it.remove();
-                continue;
-            }
-            sr.setResultObject( idMap.get( sr.getResultId() ) );
-        }
-
-        timer.stop();
-
-        if ( timer.getTime() > 200 ) {
-            BaseFormController.log.info( "Value object conversion for " + ids.size() + " " + entityClass + " after search took " + timer.getTime() + " ms." );
-        }
-    }
-
-    private Collection<ArrayDesignValueObject> filterAD( final Collection<ArrayDesignValueObject> toFilter,
-            SearchSettings settings ) {
-        // Note: if possible we should move filtering into the search service (as is done for EEs) (this is not a big deal)
-        Taxon tax = settings.getTaxon();
-        if ( tax == null )
-            return toFilter;
-        Collection<ArrayDesignValueObject> filtered = new HashSet<>();
-        for ( ArrayDesignValueObject aavo : toFilter ) {
-            if ( ( aavo.getTaxon() == null ) || ( aavo.getTaxon().equalsIgnoreCase( tax.getCommonName() ) ) ) {
-                filtered.add( aavo );
-            }
-        }
-
-        return filtered;
-    }
-
```

Comment:
This is the big chunk of logic we get rid of by centralizing the VO conversion logic.